src/operators/mod.rs

```diff
@@ -279,173 +326,234 @@ where
+                        // Output a final state update output per
+                        // epoch. Remove will ensure we slowly drain
+                        // the buffer.
+                        if is_closed(&epoch) {
+                            if let Some(key_updates) =
+                                outgoing_epoch_to_key_updates_buffer.remove(&epoch)
+                            {
+                                let mut state_backup_session =
+                                    state_backup_handle.session(&state_backup_cap);
+
+                                let backups = key_updates.into_iter().map(|(key, update)| {
+                                    StateBackup(
+                                        RecoveryKey(step_id.clone(), key, epoch.clone()),
+                                        update,
+                                    )
+                                });
 
-                                outgoing_epoch_to_key_updates_buffer
-                                    .entry(epoch.clone())
-                                    .or_insert_with(HashMap::new)
-                                    .insert(key, state_update);
+                                state_backup_session.give_iterator(backups);
                             }
                         }
-                    },
-                );
+                    }
 
-                // Then output final state update.
-                state_backup_fncater.for_each(
-                    &[input_handle.frontier(), state_loading_handle.frontier()],
-                    |state_update_cap, _ncater| {
-                        let epoch = state_update_cap.time();
-
-                        let mut state_backup_session =
-                            state_backup_handle.session(&state_update_cap);
-
-                        // Now that this epoch is complete, write out any
-                        // modified state. `remove()` will ensure we
-                        // slowly drain the buffer.
-                        if let Some(key_updates) =
-                            outgoing_epoch_to_key_updates_buffer.remove(epoch)
-                        {
-                            let backups = key_updates.into_iter().map(|(key, update)| {
-                                StateBackup(
-                                    RecoveryKey(step_id.clone(), key, epoch.clone()),
-                                    update,
-                                )
-                            });
-                            state_backup_session.give_iterator(backups);
-                        }
-                    },
-                );
+                    // Schedule an activation at the next requested
+                    // wake up time.
+                    let now = Instant::now();
+                    if let Some(delay) = key_to_next_activate_at_buffer.values().map(|a| a.duration_since(now)).min() {
+                        activator.activate_after(delay);
```

Comment:
> When we call `activate_after`, are we still only able to activate once `step` is called again?

I'm not sure if an operator can be activated multiple times on a single `step` call if the activation timeout passes while doing other work. Let me look into that. Although in our "source operators are inputs" future, I think that sidesteps this issue.

I think the main thing that could interact here is if we have an input source that is like "read data from Kafka but timeout after 1 sec". So then you could have the case where you read a ton of data but now the input source is empty so you get stuck waiting for new data from the input source when you actually have work to do downstream and that delays data entering a windowing operator resulting in windowing artifacts (like maybe random empty windows). I think the way around this is to design our input to never have blocking timeouts; if there's no useful work to do, yield back to Timely with a "wake me up later". This is a bummer in that we can't tap into a full async framework for this, but we'll see how bad it is to just have a delay wait. Also, if you're using event time, this isn't a problem.

> Is there a minimum time we should allow for windowing, based on the "heartbeat" of epochs flowing through the dataflow?

Windowing interacts with our "step while the probe is before the current input epoch" by introducing the input-to-probe processing latency as a delay between each epoch: e.g. when the input ticks over to epoch 5, the input is stalled until we see the last items of epoch 4 exit the output. That latency is not directly related to the heartbeat epoch length. But it would introduce some artifacts in windowing around that timescale if you're using a system clock to window. Again, if you were using event time, you wouldn't get those artifacts, though. And since this code is generic over clock type, I don't know we should enforce that here. Also since input-to-probe latency is an emergent property, it'd be hard to check this.